File: app/src/main/java/java238/widgets/AmodeStack.java
package java238.widgets;

import ch.bailu.gtk.adw.StatusPage;
import ch.bailu.gtk.gio.SimpleAction;
import ch.bailu.gtk.gio.SimpleActionGroup;
import ch.bailu.gtk.gtk.Button;
import ch.bailu.gtk.gtk.Stack;
import java238.App;
import java238.background.Amode;
import java238.background.AmodeList;
import java238.background.CommandInfo;

import java.util.ArrayList;
import java.util.Map;
import java.util.TreeMap;
import java.util.function.Supplier;

public class AmodeStack extends Stack {

    AmodeList amodeList;
    StatusPage noneSelected;
    Map<String, Supplier<AmodeEditorWidget>> editorWidgetSupplier = new TreeMap<>();
    public AmodeStack() {
        super();

        noneSelected = new StatusPage();
        SimpleAction action = new SimpleAction("move-up", null);
        action.onActivate((stuff) -> {
            System.out.println("asjdasdsda");
        });
        setName("amodestack");
        SimpleActionGroup group = new SimpleActionGroup();
        insertActionGroup("modestack", group.asActionGroup());
        group.asActionMap().addAction(action.asAction());
        noneSelected.setIconName("system-search-symbolic");
        noneSelected.setTitle("No Auto Selected");
        noneSelected.setDescription("Please pick an auto or load a robot project");
        addNamed(noneSelected, "none");
        setVisibleChildName("none");
        var up = Button.newFromIconNameButton("go-up-symbolic");
        App.header.packEnd(up);
        up.onClicked(this::moveFocusedCommandUp);

    }

    public AmodeList getAmodeList() {
        return amodeList;
    }

    public void setAmodeList(AmodeList list) {
        amodeList = list;
    }

    public AmodeList getUpdatedAmodesList() {
        ArrayList<Amode> modesList = new ArrayList<>(editorWidgetSupplier.size());
        for (String key : editorWidgetSupplier.keySet()) {
            modesList.add(editorWidgetSupplier.get(key).get().getUpdatedMode());
        }



        amodeList.setAutonomousModes(modesList);
        return amodeList;
    }

    public void renameMode(String oldName, String newName) {

    }

    public void removeAmode() {
        editorWidgetSupplier.remove(getVisibleChildName().toString());
        remove(getVisibleChild());
        setVisibleChildName("none");
    }

    public void addCommandToVisibleChild(CommandInfo info) {
        System.out.println(getVisibleChild().getName().toString());
        editorWidgetSupplier.get(getVisibleChild().getName().toString()).get().addCommand(info);
    }

    public void addTitled(AmodeEditorWidget widget, String name) {
        super.addTitled(widget, name, name);
        editorWidgetSupplier.put(name, widget::getSelf);
    }

    public void addAmodeToList() {
        Amode newMode = new Amode();
        newMode.setName("New Auto");
        AmodeEditorWidget editor = new AmodeEditorWidget(newMode);
        addTitled(editor, "New Auto");
    }

    public void moveFocusedCommandUp() {
        editorWidgetSupplier.get(getVisibleChild().getName().toString()).get().moveRowUp();
        System.out.println("WOAH");
    }
}

File: app/src/main/java/java238/widgets/AmodeStack.kt
package java238.widgets

import ch.bailu.gtk.adw.StatusPage
import ch.bailu.gtk.gio.SimpleAction
import ch.bailu.gtk.gio.SimpleActionGroup
import ch.bailu.gtk.glib.Variant
import ch.bailu.gtk.gobject.ParamSpec
import ch.bailu.gtk.gtk.Button
import ch.bailu.gtk.gtk.Entry
import ch.bailu.gtk.gtk.Popover
import ch.bailu.gtk.gtk.Stack
import ch.bailu.gtk.gtk.StackPage
import ch.bailu.gtk.gtk.Widget
import ch.bailu.gtk.type.PointerContainer
import ch.bailu.gtk.type.Str
import java238.App
import java238.background.Amode
import java238.background.AmodeList
import java238.background.CommandInfo
import java.util.*
import java.util.function.Supplier

class AmodeStack : Stack() {
    var amodeList: AmodeList? = null
    var noneSelected: StatusPage
    var editorWidgetSupplier: MutableMap<String, Supplier<AmodeEditorWidget>> = TreeMap()
    var up: Button
    var down: Button

    init {
        noneSelected = StatusPage()
        val action = SimpleAction("move-up", null)
        action.onActivate { println("asjdasdsda") }
        setName("amodestack")
        val group = SimpleActionGroup()
        insertActionGroup("modestack", group.asActionGroup())
        group.asActionMap().addAction(action.asAction())
        noneSelected.setIconName("system-search-symbolic")
        noneSelected.setTitle("No Auto Selected")
        noneSelected.setDescription("Please pick an auto or load a robot project")
        addNamed(noneSelected, "none")
        setVisibleChildName("none")
        up = Button.newFromIconNameButton("go-up-symbolic")
        up.onClicked { moveFocusedCommandUp() }
        down = Button.newFromIconNameButton("go-down-symbolic")
        down.onClicked { moveFocusedCommandDown() }

        val rename = Button.newFromIconNameButton("document-edit-symbolic").also {
            it.onClicked {renameMode()}
            App.header.packStart(it)

        }

        onNotify { signal: ParamSpec ->
            if (signal.name.toString() == "visible-child") {
                App.title.subtitle = visibleChildName
            }
        }
    }

    private fun moveFocusedCommandDown() {
        editorWidgetSupplier[visibleChild.name.toString()]!!.get().moveRowDown()
    }

    val updatedAmodesList: AmodeList?
        get() {
            val modesList = ArrayList<Amode>(editorWidgetSupplier.size)
            for (key in editorWidgetSupplier.keys) {
                modesList.add(editorWidgetSupplier[key]!!.get().updatedMode)
            }
            amodeList!!.setAutonomousModes(modesList)
            return amodeList
        }

    private fun renameMode() {
        val namePop = Popover()
        val entry = Entry()
        namePop.child = entry
        namePop.parent = App.title
        entry.onActivate {
            val supp = editorWidgetSupplier.run {

                getPage(visibleChild).title = entry.asEditable().text.also { getPage(visibleChild).name = it }
                put(entry.asEditable().text.toString(), remove(visibleChild.name.toString())!!.also { it.get().name = visibleChildName })
                App.title.subtitle = entry.asEditable().text
                namePop.hide()
            }
        }
        entry.asEditable().text = visibleChildName
        namePop.popdown()
        namePop.show()
        println("woah")
    }
    fun removeAmode() {
        editorWidgetSupplier.remove(visibleChildName.toString())
        getPage(visibleChild).run {
            title = Str("")
            this.visible= false
        }

    }

    fun addCommandToVisibleChild(info: CommandInfo?) {
        println(visibleChild.name.toString())
        editorWidgetSupplier[visibleChild.name.toString()]!!.get().addCommand(info!!)
    }

    fun addTitled(widget: AmodeEditorWidget?, name: String) {
        super.addTitled(widget!!, name, name)
        editorWidgetSupplier[name] = Supplier { widget.self }
    }

    fun addAmodeToList() {
        val newMode = Amode()
        newMode.name = "New Auto"
        val editor = AmodeEditorWidget(newMode)
        addTitled(editor, "New Auto")
    }

    fun moveFocusedCommandUp() {
        editorWidgetSupplier[visibleChild.name.toString()]!!.get().moveRowUp()
        println("WOAH")
    }
}
